packages/gatsby-image/src/index.js

```diff
@@ -107,44 +107,60 @@ function getIO() {
   return io
 }
 
+// Return an array ordered by elements having a media prop, does not use native sort, as a stable sort is not guaranteed by all browsers/versions
+function groupByMedia(imageVariants) {
+  const withMedia = []
+  const without = []
+  imageVariants.forEach(variant =>
+    (variant.media ? withMedia : without).push(variant)
+  )
+
+  return [...withMedia, ...without]
+}
+
 function generateImageSources(imageVariants) {
-  return imageVariants.map(({ src, srcSet, srcSetWebp, media, sizes }) => (
-    <React.Fragment key={src}>
-      {srcSetWebp && (
-        <source
-          type="image/webp"
-          media={media}
-          srcSet={srcSetWebp}
-          sizes={sizes}
-        />
-      )}
-      <source media={media} srcSet={srcSet} sizes={sizes} />
-    </React.Fragment>
-  ))
+  return groupByMedia(imageVariants).map(
+    ({ src, srcSet, srcSetWebp, media, sizes }) => (
+      <React.Fragment key={src}>
+        {srcSetWebp && (
+          <source
+            type="image/webp"
+            media={media}
+            srcSet={srcSetWebp}
+            sizes={sizes}
+          />
+        )}
+        <source media={media} srcSet={srcSet} sizes={sizes} />
+      </React.Fragment>
+    )
+  )
 }
 
 function generateTracedSVGSources(imageVariants) {
-  return imageVariants.map(({ src, media, tracedSVG }) => (
+  return groupByMedia(imageVariants).map(({ src, media, tracedSVG }) => (
```

Comment:
This wasn't in my pastebin, the `groupByMedia` call should only be run once in `render()` not in any of these generate methods.